search/reverse_geocoder.cpp

```diff
@@ -89,88 +101,74 @@ size_t ReverseGeocoder::GetMatchedStreetIndex(string const & keyName,
   return result;
 }
 
+pair<vector<ReverseGeocoder::Street>, uint32_t>
+ReverseGeocoder::GetNearbyFeatureStreets(FeatureType const & ft) const
+{
+  pair<vector<ReverseGeocoder::Street>, uint32_t> result;
+
+  GetNearbyStreets(const_cast<FeatureType &>(ft), result.first);
+
+  HouseTable table;
+  if (!table.Get(m_index, ft.GetID(), result.first, result.second))
+    result.second = numeric_limits<uint32_t>::max();
+
+  return result;
+}
+
 void ReverseGeocoder::GetNearbyAddress(m2::PointD const & center, Address & addr) const
 {
   vector<Building> buildings;
   GetNearbyBuildings(center, buildings);
 
-  vector<Street> streets;
-  unique_ptr<search::v2::HouseToStreetTable> table;
-  MwmSet::MwmHandle mwmHandle;
+  HouseTable table;
+  int triesCount = 0;
 
   for (auto const & b : buildings)
   {
-    if (!table || mwmHandle.GetId() != b.m_id.m_mwmId)
-    {
-      mwmHandle = m_index.GetMwmHandleById(b.m_id.m_mwmId);
-      if (!mwmHandle.IsAlive())
-        continue;
-      table = search::v2::HouseToStreetTable::Load(*mwmHandle.GetValue<MwmValue>());
-    }
-
-    GetNearbyStreets(b.m_center, streets);
-
-    uint32_t ind;
-    if (table->Get(b.m_id.m_index, ind) && ind < streets.size())
-    {
-      addr.m_building = b;
-      addr.m_street = streets[ind];
-      return;
-    }
+    if (GetNearbyAddress(table, b, addr) || (++triesCount == 5))
+      break;
   }
 }
 
-pair<vector<ReverseGeocoder::Street>, uint32_t> ReverseGeocoder::GetNearbyFeatureStreets(
-    FeatureType const & feature) const
+void ReverseGeocoder::GetNearbyAddress(FeatureType & ft, Address & addr) const
 {
-  pair<vector<ReverseGeocoder::Street>, uint32_t> result;
-  auto & streetIndex = result.second;
-  streetIndex = numeric_limits<uint32_t>::max();
+  HouseTable table;
+  (void)GetNearbyAddress(table, FromFeature(ft, 0.0 /* distMeters */), addr);
+}
 
-  FeatureID const fid = feature.GetID();
-  MwmSet::MwmHandle const mwmHandle = m_index.GetMwmHandleById(fid.m_mwmId);
-  if (!mwmHandle.IsAlive())
+bool ReverseGeocoder::GetNearbyAddress(HouseTable & table, Building const & bld,
+                                       Address & addr) const
+{
+  string street;
+  if (osm::Editor::Instance().GetEditedFeatureStreet(bld.m_id, street))
   {
-    LOG(LERROR, ("Feature handle is not alive", feature));
-    return result;
+    addr.m_building = bld;
+    addr.m_street.m_name = street;
+    return true;
   }
 
-  auto & streets = result.first;
-  GetNearbyStreets(feature::GetCenter(feature), streets);
+  vector<Street> streets;
+  GetNearbyStreets(bld.m_id.m_mwmId, bld.m_center, streets);
 
-  unique_ptr<search::v2::HouseToStreetTable> const table =
-      search::v2::HouseToStreetTable::Load(*mwmHandle.GetValue<MwmValue>());
+  uint32_t ind;
+  if (table.Get(m_index, bld.m_id, streets, ind))
```

Comment:
As you always pass m_index as an argument, please, move it to the table constructor.